### PR TITLE
Add TFM check for `Publish-Trimmed-ReadyToRun-SingleFile`

### LINKED ISSUE #3728: How to debug "GenerateBundle" task failed unexpectedly
# Issue Title

dotnet publish xxx.csproj -r win-x64 -c Release /p:PublishSingleFile=true

gives me the following:

C:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(809,5): error MSB4018: The "GenerateBundle" task failed unexpectedly. [C:\xxx.csproj]
C:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(809,5): error MSB4018: Microsoft.NET.HostModel.AppHost.PlaceHolderNotFoundInAppHostException: Exception of type 'Microsoft.NET.HostModel.AppHost.PlaceHolderNotFoundInAppHostException' was thrown. [C:\xxx.csproj]
C:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(809,5): error MSB4018:    at Microsoft.NET.HostModel.AppHost.BinaryUtils.SearchAndReplace(MemoryMappedViewAccessor accessor, Byte[] searchPattern, Byte[] patternToReplace, Boolean pad0s) [C:\xxx.csproj]
C:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(809,5): error MSB4018:    at Microsoft.NET.HostModel.AppHost.BinaryUtils.SearchAndReplace(String filePath, Byte[] searchPattern, Byte[] patternToReplace, Boolean pad0s) [C:\xxx.csproj]
C:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(809,5): error MSB4018:    at Microsoft.NET.HostModel.AppHost.HostWriter.<>c__DisplayClass3_0.<SetAsBundle>b__0() [C:\xxx.csproj]
C:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(809,5): error MSB4018:    at Microsoft.NET.HostModel.RetryUtil.RetryOnIOError(Action func) [C:\xxx.csproj]
C:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(809,5): error MSB4018:    at Microsoft.NET.HostModel.AppHost.HostWriter.SetAsBundle(String appHostPath, Int64 bundleHeaderOffset) [C:\xxx.csproj]
C:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(809,5): error MSB4018:    at Microsoft.NET.HostModel.Bundle.Bundler.GenerateBundle(IReadOnlyList`1 fileSpecs) [C:\xxx.csproj]
C:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(809,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateBundle.ExecuteCore() [C:\xxx.csproj]
C:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(809,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [C:\xxx.csproj]
C:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(809,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [C:\xxx.csproj]
C:\Program Files\dotnet\sdk\3.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Publish.targets(809,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [C:\xxx.csproj]

# General

Would love to be able to provide more info, but how?

Is there a debug publish switch?

### PR DESCRIPTION
**Pull request by swaroop-sridhar**
Trying to publish a `.net core 2.1` app as a single file fails with the `PlaceHolderNotFoundInAppHostException` generated by the `HostWriter`.

Publishing apps as a single file is only supported in `.net core 3` and higher.
So, this change adds a version check in the SDK.

Similarly, the check is also added for the `PublishReadyToRun` and `PublishTrimmed` options.

Fixes https://github.com/dotnet/sdk/issues/3728

### CODE REVIEW

**Comment by swaroop-sridhar** [comment]
CC: @nguerrera  please review.

**Comment by swaroop-sridhar** [comment]
CC: @fadimounir @sbomer  @vitek-karas

**Comment by fadimounir** [comment]
LGTM

**Comment by fadimounir** [comment]
BTW, we shouldn't be executing R2R for any version less than 3.0 today.
see: https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets#L214

**Comment by swaroop-sridhar** [comment]
@fadimounir Shouldn't we check for these conditions, and fail ... rather than silently not generate R2R code?

**Comment by fadimounir** [comment]
Yes it's better to either fail, or emit a warning that it's not supported and skip the R2R/Trimming/SingleExe step. I'm good with either options

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets [hunk 1]
```diff
@@ -142,6 +142,15 @@ Copyright (c) .NET Foundation. All rights reserved.
...
 
+    <NETSdkError Condition="'$(PublishSingleFile)' == 'true' and '$(_TargetFrameworkVersionWithoutV)' &lt; '3.0'"
+                 ResourceName="PublishSingleFileRequiresVersion30" />
+
+    <NETSdkError Condition="'$(PublishReadyToRun)' == 'true' and '$(_TargetFrameworkVersionWithoutV)' &lt; '3.0'"
+                 ResourceName="PublishReadyToRunRequiresVersion30" />
+
+    <NETSdkError Condition="'$(PublishTrimmed)' == 'true' and '$(_TargetFrameworkVersionWithoutV)' &lt; '3.0'"
+             ResourceName="PublishTrimmedRequiresVersion30" />
+
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets]: These should also check TargetFrameworkIdentifier is .NETCoreApp.

**Comment by nguerrera** [comment]
What release do you want this in? Should it sit until 5 or go out with 3.1 soon? I think 3.1. Can you rebase and retarget to release/3.1.1xx? We don't dual checkin master and 3.1, we rely on merge from 3.1 to master.

**Comment by swaroop-sridhar** [comment]
OK Thanks @nguerrera, I'll put the PR for 3.1xx branch

**Comment by nguerrera** [comment]
For R2R and ILLink, this is a breaking change. They both no-op now on less than 3.0.

I think this is better, but it makes me re-question a bit if it should go in 3.1.

@livarcocc @dsplaisted What do you think?

One minor scenario would be if you multi-target and want to use these with your 3.0+ targets, but not the others. You would now have to add a condition on TargetFramework. This is arguably more obvious to reader of project file what would happen, but it is breaking.

**Comment by swaroop-sridhar** [comment]
@nguerrera, how about making it a warning for 3.0 for R2R and Trimming in 3.1? 

Also, how about adding warnings (for R2R and Trimming) if the following checks fail? 
`'$(TargetFrameworkIdentifier)' == '.NETCoreApp'` ` '$(_IsExecutable)' == 'true'"` 
That is, emitting a warning, rather than silently failing to comply with the option? 

For R2R and IL Trimming, I think the above checks should only generate warnings, because they may be accomodated in future versions (ex: R2R code for libs).

**Comment by nguerrera** [comment]
Replaced by #3766

**Pull request closed without merging**